### PR TITLE
Add dict(*iterable) constructor

### PR DESCRIPTION
**Pull request by guilhermeleobas**
Fix issue #8328

### CODE REVIEW

File: numba/typed/dictimpl.py
Comment (esc) [inline numba/typed/dictimpl.py:15]:
```suggestion
def dict_constructor(context, builder, sig, args):
```

File: numba/typed/dictimpl.py
Comment (esc) [inline numba/typed/dictimpl.py:21]:
```suggestion
    def dict_impl(iterable):
```

File: numba/typed/dictimpl.py
Comment (esc) [inline numba/typed/dictimpl.py:27]:
```suggestion
    return context.compile_internal(builder, dict_impl, sig, args)
```

**Review by esc — CHANGES_REQUESTED** [review]
I gave this a first pass and found some potential copy and paste errors.

File: numba/typed/dictimpl.py
Comment (guilhermeleobas) [reply numba/typed/dictimpl.py]: Fixed!

File: numba/core/typing/dictdecl.py
Comment (stuartarchibald) [inline numba/core/typing/dictdecl.py:20]:
```suggestion
The only supported uses are `dict()` or `dict(iterable)`.
```

File: docs/source/reference/pysupported.rst
Comment (stuartarchibald) [inline docs/source/reference/pysupported.rst:730]:
```suggestion
Numba also supports, explicitly, the ``dict(iterable)`` constructor.
```

File: numba/tests/test_dictobject.py
Comment (stuartarchibald) [inline numba/tests/test_dictobject.py:1194]: This base class will run as a unit test, is that intended?

File: numba/tests/test_dicts.py
Comment (stuartarchibald) [inline numba/tests/test_dicts.py:42]:
```suggestion
        # Test dict(iterable)
```

File: numba/typed/typeddict.py
Comment (stuartarchibald) [inline numba/typed/typeddict.py:136]: This behaviour isn't working under JIT?

File: numba/tests/test_dictobject.py
Comment (stuartarchibald) [inline numba/tests/test_dictobject.py:1316]: Here's some things that either aren't working or could work better:
```python
from numba import njit
import numpy as np

# Valid, won't compile, fails at typing.
@njit
def foo():
    return dict(([1, 2], [1, 2]))

print(foo())


#Wrong error message, the issue is the types in the iterable aren't supported
#whereas the error message complains of: `Unsupported use of `dict()` with
#positional or keyword argument(s). The only supported uses are `dict()` or
#`dict(*iterable)`.`
@njit
def foo():
    return dict(((1, 2), (3, 'a')))

print(foo())


#Lowering error, should be caught earlier
@njit
def foo():
    return dict((('a', 'b', 'c'), ('d', 'e', 'f')))

print(foo())


#This should really complain about unhashable key type
@njit
def foo():
    return dict((('a', 1), ({}, 2)))

print(foo())


# Mapping support not working under JIT
@njit
def foo():
    return dict({x: x ** 2 for x in range(3)})

print(foo())
```

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch @guilhermeleobas, I've not looked at the code in depth yet, but have provided some feedback about things that might need looking at in terms of usability and correctness. Thanks again!

File: numba/typed/typeddict.py
Comment (guilhermeleobas) [reply numba/typed/typeddict.py]: I'll leave `dict(mapping)` ctor for a future PR and focus exclusively on `dict(iterable)` in this one.

File: numba/tests/test_dictobject.py
Comment (guilhermeleobas) [reply numba/tests/test_dictobject.py]: I believe I have address all the issues pointed out in our comment, except for the last one, which involves `dict(mapping)`.

**Comment by guilhermeleobas** [comment]
Hi @stuartarchibald, I believe I have addressed all your comments. The current failure is not related to any changes introduced in my previous commit.

```
- Windows py310_np122

(testenv) D:\a\1\s>cmd /C conda install -q -y -c numba tbb=2021 tbb-devel 
Collecting package metadata (current_repodata.json): ...working... failed

CondaHTTPError: HTTP 504 GATEWAY TIMEOUT for url <[https://conda.anaconda.org/numba/win-64/repodata.json>](https://conda.anaconda.org/numba/win-64/repodata.json%3E)
Elapsed: 01:00.076791
CF-RAY: 73f5e5f559ed2ad6-ORD

A remote server error occurred when trying to retrieve this URL.
```

**Comment by esc** [comment]
@guilhermeleobas @kc611 -- may kindly request an update on the state of this PR?

**Comment by guilhermeleobas** [comment]
Hi @esc, PR is waiting for another review

File: numba/core/typing/dictdecl.py
Comment (kc611) [inline numba/core/typing/dictdecl.py:33]: One question:

```python
@njit
def foo():
    return dict(([1, 2], [1, 2])) # Error

print(foo())
```

This is valid within CPython, But this PR only intends to add support when arguments are a list or a tuple of tuples (and not the other cases). Is that correct ?

**Review by kc611 — APPROVED** [review]
I gave it a review, its LGTM (so far). A question below.

File: numba/core/typing/dictdecl.py
Comment (guilhermeleobas) [reply numba/core/typing/dictdecl.py:33]: If possible, I'd like to leave this use case for a future PR.

**Review by kc611 — APPROVED** [review]

**Review by sklam — APPROVED** [review]

**Pull request merged**
